rfcs/20220217-visual_regression_testing.md

```diff
@@ -0,0 +1,240 @@
+
+Ultimately I think the best option would be to regularly (once a day) run a scheduled maintenance task to remove any historical snapshots older than a month that are not currently used by the main branch. That would mean we would always have a _reasonable_ historical view of snapshots that change frequently and I frankly doubt snapshot history extending more than a month back would be very useful anyway.
+
+## Cross platform rendering
+
+Browsers will sometimes render minute differences between platforms, even with the exact same webpage. This means that contributors running Linux might generate different screenshots from those running macOS and again from those running Windows, even if they're all using the same browser and version.
+
+To avoid this, other visual regression testing libraries like BackstopJS support running tests inside a docker container and using the browser inside the container. We can do the same with Playwright (it's just not as easy).
+
+I was about to figure out "a way" of doing it. Here's what I came up with:
+
+1. Create a `./bin/visual-regression.js` file with the following:
+```js
+/**
+ * Run this as a node script so that we can retrieve the `process.cwd`
+ * which isn't possible to evaluate cross-platform inside an package.json script
+ *
+ * We can also automatically sync the local playwright version
+ * with the docker image version. This ensures that the browser
+ * versions expected by the `playwright` binary installed locally
+ * matches the versions pre-installed in the container.
+ */
+const { spawnSync } = require('child_process')
+const fs = require('fs')
+const path = require('path')
+const yaml = require('yaml')
+
+const pnpmLock = yaml.parse(
+  String(fs.readFileSync(path.resolve(process.cwd(), 'pnpm-lock.yaml')))
+)
+
+const playwrightVersion = pnpmLock.devDependencies['@playwright/test']
+
+const args = [
+  'run',
+  '--rm',
+  '-it',
+  '--mount',
+  `type=bind,source=${process.cwd()},target=/src`,
+  '--add-host=host.docker.internal:host-gateway', // This is necessary to make `host.docker.internal` work on Linux. Does it break the command for other OSs?
+  `mcr.microsoft.com/playwright:v${playwrightVersion}-focal`,
+  '/src/bin/visual-regression.sh',
+]
+
+if (process.argv.includes('-u')) {
+  args.push('-u')
+}
+
+spawnSync('docker', args, {
+  stdio: 'inherit',
+})
```

Comment:
I believe this line is missing the 3 trailing backtics and is causing the rendering to get messed up (I'd put up a "suggestion" it but the markdown is not happy about it lol).

Reply:
Fixed! BTW to be able to use backticks inside of `` `inline code text` `` or ```` ```multi-line suggestions``` ```` you have to use an extra back tick on the fence than what you're using inside the block.

For example:

``` `` `inline code text` `` ```

Will render:

`` `inline code text` ``

And for multi-line it's the same:

`````
````
```py
pring("Hello, world!")
```
````
`````

Will render

````
```py
print("Hello, world!")
```
````

Might be easier to edit this comment and inspect the markdown content of it if it is still confusing :slightly_smiling_face:

Reply:
TIL!! This rocks!